crates/runtime/src/dataconnector.rs

```diff
@@ -247,61 +239,108 @@ pub async fn register_connector_factory(
 pub async fn create_new_connector(
     name: &str,
     params: HashMap<String, SecretString>,
+    secrets: Arc<RwLock<Secrets>>,
 ) -> Option<AnyErrorResult<Arc<dyn DataConnector>>> {
     let guard = DATA_CONNECTOR_FACTORY_REGISTRY.lock().await;
 
     let connector_factory = guard.get(name);
 
     match connector_factory {
-        Some(factory) => Some(factory(params).await),
+        Some(factory) => {
+            let mut params = remove_prefix_from_hashmap_keys(params, factory.prefix());
+            let secret_guard = secrets.read().await;
+
+            // Try to autoload secrets that might be missing from params.
+            for secret_key in factory.autoload_secrets().iter().copied() {
+                let secret_key_with_prefix = format!("{}_{secret_key}", factory.prefix());
+                tracing::debug!(
+                    "Attempting to autoload secret for {name}: {secret_key_with_prefix}",
+                );
+                if params.contains_key(secret_key) {
+                    continue;
+                }
+                let secret = secret_guard.get_secret(&secret_key_with_prefix).await;
+                if let Ok(Some(secret)) = secret {
+                    tracing::debug!("Autoloading secret for {name}: {secret_key_with_prefix}",);
+                    // Insert without the prefix into the params
+                    params.insert(secret_key.to_string(), secret);
+                }
+            }
+            let result = factory.create(params).await;
+            Some(result)
+        }
         None => None,
     }
 }
 
 pub async fn register_all() {
-    register_connector_factory("localhost", localhost::LocalhostConnector::create).await;
+    register_connector_factory("localhost", localhost::LocalhostConnectorFactory::new_arc()).await;
```

Comment:
Instead of registering a function that returns a data connector, we now register an `Arc<dyn DataConnectorFactory>` that has methods to create a data connector, but also get the prefix and autoload secrets